frb_example/dart_minimal/rust/src/api/minimal.rs

```diff
@@ -1,10 +1,37 @@
-use flutter_rust_bridge::frb;
+use flutter_rust_bridge::*;
+
+enable_frb_logging!();
+// TODO remove once unit tests are written
+// enable_frb_logging!(
+//     name = "FRBLogger",
+//     maxLoglevel = "INFO",
+//     customLogFunction = (|record: Log2DartLogRecord| {
+//         let timestamp = chrono::Local::now();
+//         let max_log_level = from_u16(record.level_number);
+//         let lang = if record.rust_log { "Rust" } else { "Dart" };
+//         let logger_name = record.logger_name;
+//         let message = record.message;
+//         println!("[{timestamp:?} {max_log_level} @{lang} {logger_name}] {message})");
+//     })
+// );
+
+// enable_frb_logging!(name = "ParamAloneLogger");
+// enable_frb_logging!(name = "ParamInfoLogger", maxLoglevel = "warning");
+// enable_frb_logging!(maxLoglevel = "warning");
+// enable_frb_logging!(
+//     customLogFunction = (|record: Log2DartLogRecord| {
+//         let logger_name = record.logger_name;
+//         let message = record.message;
+//         println!("[{logger_name}] {message})");
+//     })
+// );
 
 #[frb(init)]
```

Comment:
nit: thinking about what should be exposed to end user, i.e. what should be the code end user put/see in minimal.rs

### choice 1

```rust
flutter_rust_bridge::setup_default_logging!();

#[frb(init)]
pub fn init_app() {
    flutter_rust_bridge::setup_default_user_utils();
}
```

### choice 2

```rust
flutter_rust_bridge::setup_default_user_utils!();
```

where `setup_default_user_utils`'s definition is, "call setup_default_logging + that fn init_app".
and we still allow customizations. For example, suppose one wants to remove logging, but keep others, then he or she can remove setup_default_user_utils, and copy-paste the `fn initapp` or whatever he likes.

pros: less confusing to users (in choice 1, users may think "why logging is not user utils?")

pros: less verbose (in choice 1, user may think "oh so many lines")

cons: a bit less explicit. (but users can click that macro to see what is going on, so maybe acceptable)

Reply:
I agree with choice 2 as well :) probably there will be more supporting tools in future - and then it would get very (too) verbose. So hiding that all behind something like `setup_default_(...)` sounds very good to me :)

I am quite confident, the I can rewrite the code so that the customizing can be done any time ... from setting the log level to a custom log function.

I think this would work from both sides - meaning one could call a Rust function `set_custom_log_fn` from `minimal.rs` or `main.dart`. So my next question is not very relevant, but: We are assuming a user would do the logging configuration preferably from Dart, right? 

Thus, for setting the log level one needs to use the Dart logging package's definition (which I find quite unusual ...), even when setting it from Rust. I could easily change this though, I was aiming for consistency.

As for the custom log function, if set from Dart it shall be Dart code, if set from Rust Rust code. I think we can change this as well, but that requires a bit more effort. But probably this is a good default.

So, to summarize, I think it would be best if:
1. logging gets automatically initialized with defaults, triggered as part of RustLib.init()
2. thus, there is no visible code in neither `main.dart`, nor `minimal.rs`
3. issuing log statements works as usual in the respective language (`Logger().info('from Dart)`, log::info(`from Rust`) - this should already work :)
4. Customizing the log level can be done from both, Dart or Rust code
5. the generated code ends up in frb_generated.rs/dart and not minimal.rs

For (5.), I wasn't able to get it there - the StreamSink was never processed. I tried calling the macro in `frb_rust/src/for_generated/boilerplate.rs`, `frb_rust/src/misc/user_utils.rs`, etc. Probably you see right away what needs to be fixed :)

Reply:
> probably there will be more supporting tools in future

Totally agree. Choice 1 will require user to migrate again at that time which is not optimal.

Agree 1-6! Except minor details:

> A user would disable this by not having the log feature active

I guess by simply copy-paste content in setup_default_user_utilities and remove the one about logging.
Using feature for this, personally speaking, may be a little bit overkill, and a bit less flexible.

> Customizing the log level can be done from both, Dart or Rust code

Yes, for simplicity (of our implementation), users can use Rust and Dart's native mechanism to set logging level respectively. 

> the generated code ends up in frb_generated.rs/dart and not minimal.rs

I will handle that

Reply:
> > A user would disable this by not having the log feature active
> 
> I guess by simply copy-paste content in setup_default_user_utilities and remove the one about logging. Using feature for this, personally speaking, may be a little bit overkill, and a bit less flexible.

Yes, true, that would be a way as well. I was referring to the already existing feature flag 'log' - I used `#[cfg(feature = "log")]` wherever it made sense :)


> > Customizing the log level can be done from both, Dart or Rust code
> 
> Yes, for simplicity (of our implementation), users can use Rust and Dart's native mechanism to set logging level respectively.

Oh yes, we can just use the usual functions for setting the log level - will do that.
 
> > the generated code ends up in frb_generated.rs/dart and not minimal.rs
> 
> I will handle that

Ah, great , thanks :)

Reply:
That flag was there because I wanted frb_rust to have minimal dependency if a user wants it. It accidentally also controls that existing behavior...

Reply:
Good idea :) Ah, ok. If this flag is needed always for the existing behavior (mandatory parts?) I would suggest to remove it from these and repurpose it for this logging implementation. Probably users have used it for that. I actually assume that probably nobody deactivated it ... as this is not so straight-forward in Rust.